### ISSUE TITLE
semi-supervised learning for cell annotation

### ISSUE THREAD

**Issue by jeff-regier** [YosefLab/scVI-dev#59]
Maybe start with one fully annotated dataset but reserve most of the annotations/labels for testing. Show that we can recover the correct labels with semi-supervised ladder networks. 

https://arxiv.org/pdf/1507.02672.pdf

Then add a second dataset---or maybe just a second batch of the same dataset. Start incorporating the scVI graphical model into the ladder network, to account for batch effects, using the approach in 

https://arxiv.org/pdf/1602.02282.pdf

**Comment by jeff-regier** [comment]
@Edouard360 @maxime1310  We may also want to discuss the original semi-supervised VAE:
http://papers.nips.cc/paper/5352-semi-supervised-learning-with-deep-generative-models.pdf

**Comment by Edouard360** [comment]
* Instead of concatenating simply [x, y] to get z, z_mu, z_var, I have done something much more complicated (code can be simplified)

https://github.com/YosefLab/scVI-dev/blob/c6dd6966708ea42f522ce6b9439f5a0277ae7b79/scvi/scvi.py#L200-L210

The idea was to "stack" n_labels predictors for z_mu, z_var, so that the last layer of the encoder has n_latent*n_labels output, instead of just n_latent, and then I index with the label y, to get the vector of n_latent parameters describing z. But we'll probably simply turn to the concatenation [x, y] -> z.

* Instead of creating a new Model - ( maybe using inheritance ? ), I added a flag to our currently existing implementation which makes code less readable.

https://github.com/YosefLab/scVI-dev/blob/c6dd6966708ea42f522ce6b9439f5a0277ae7b79/scvi/scvi.py#L33

* I haven't handled labelled and unlabelled data separately, but assumed they could come in any fashion ( -1 indicated unlabelled data). All things considered, this might be a bad for code clarity.

https://github.com/YosefLab/scVI-dev/blob/c6dd6966708ea42f522ce6b9439f5a0277ae7b79/scvi/train.py#L18-L19

I had forced myself not to look at this [repo](https://github.com/wohlert/semi-supervised-pytorch), which looks strong in terms of code structure and implements a lot of the things we discussed. (Ladder VAE, Normalizing Flows, ...) I guess I will take some inspiration from there to correct my mistakes and adapt the code.

**Comment by Edouard360** [comment]
@maxime1310  and I have been working on this [branch](/YosefLab/scVI-dev/tree/semi-supervised) for semi-supervised learning.

It is taking the latest changes from master, and I also changed pytorch in requirements to v0.4 and will further refactor (more in #72).

We would like **to port to pytorch 0.4 from this branch** since it has some refactoring changes that we believe are good compared to the current code in master.

**Comment by Edouard360** [comment]
Unfortunately, whether on master or on this [branch](/YosefLab/scVI-dev/tree/semi-supervised), in torch v0.4, we still get the Segmentation Fault...

**Comment by jeff-regier** [comment]
Does the seg fault happen even when num_workers is set to 0?

**Comment by Edouard360** [comment]
Actually it also happens with numpy matrices (I tested it)

**Comment by soumith** [comment]
if you're using multiprocessing, try setting (at the top of your main.py):

```
torch.multiprocessing.set_start_method("spawn")
```

Most libraries in the wild aren't built to be fork-safe, so weird side-effects like segfaults appears with default python multiprocessing.
`spawn` will create a new process memory space, no fork-semantics, so libraries that are not fork-safe will work well.
